# Pandas lab 1

## **Import data and libraries**

In this Lab we will use a restaurant dataset. 

The dataset contains tips data from different customers females and males, smokers and non smokers, from days, dinner or lunch and from different table size.

In [24]:
import pandas as pd

# Load restaurant.csv file from data folder to variable "df":
df = pd.read_csv(r'Data\restaurant.csv') # your code here
df

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [25]:
# show the first 5 rows
# your code here
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

## **Overview**

### 📑 Question 1 (EXPRESSION - 5 pts)

Write **one line of code** to Show the **first 3 rows** of the data

**Output should look like this**

|    |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|---:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
|  0 |        16.99 |  1.01 | Female | No       | Sun   | Dinner |      2 |
|  1 |        10.34 |  1.66 | Male   | No       | Sun   | Dinner |      3 |
|  2 |        21.01 |  3.5  | Male   | No       | Sun   | Dinner |      3 |

In [27]:
# Your answer/code here:
df.head(3)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3

### 📑 Question 2 (EXPRESSION - 5 pts)

Write **one line of code** to Show the **last 2 rows** of the data

**Output should look like this**

|     |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|----:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
| 242 |        17.82 |  1.75 | Male   | No       | Sat   | Dinner |      2 |
| 243 |        18.78 |  3    | Female | No       | Thur  | Dinner |      2 |

In [28]:
# Your answer/code here:
df.tail(2)

total_bill   tip     sex smoker   day    time  size
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

In [29]:
df.shape

(244, 7)

### 📑 Question 3 (MULTICHOICE_SINGLE - 5 pts)

What syntax is right to get random rows in dataframe?

-  A df.shuffle()

- B df.random()

- C df.sample()

- D df.mix()

In [31]:
# Your answer/code here:
df.sample(3)

total_bill   tip     sex smoker   day    time  size
106       20.49  4.06    Male    Yes   Sat  Dinner     2
231       15.69  3.00    Male    Yes   Sat  Dinner     3
134       18.26  3.25  Female     No  Thur   Lunch     2

### 📑 Question 4 (EXPRESSION - 5 pts)
Write one line of code to show a **descriptive statistics summary** of the dataset.

**Output should look like this**

|       |   total_bill |       tip |      size |
|:------|-------------:|----------:|----------:|
| count |    244       | 244       | 244       |
| mean  |     19.7859  |   2.99828 |   2.56967 |
| std   |      8.90241 |   1.38364 |   0.9511  |
| min   |      3.07    |   1       |   1       |
| 25%   |     13.3475  |   2       |   2       |
| 50%   |     17.795   |   2.9     |   2       |
| 75%   |     24.1275  |   3.5625  |   3       |
| max   |     50.81    |  10       |   6       |

In [33]:
# Your answer/code here:
df.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

## **EDA - Exploratory Data Analysis**

### 📑 Question 5 (EXPRESSION - 5 pts)
Write one line of code to find out the **minimum amount of tip**.

**Output should be a number**

In [34]:
# Your answer/code here:
df['tip'].min()

1.0

### 📑 Question 6 (EXPRESSION - 5 pts)
Write one line of code to find out the **maximum amount of tip**.

**Output should be a number**

In [35]:
# Your answer/code here:
df['tip'].max()

10.0

### 📑 Question 7 (MULTICHOICE_SINGLE - 5 pts)

On which days of the week, the restaurant is open?

- A. Mon, Tue, Wed, Thu

- B. Thu, Fri, Sat, Sun

- C. Mon, Wed, Fri 

- D. Open all days in week

In [36]:
# Your answer/code here: ---> Option B
print('Các ngày nhà hàng mở cửa',df['day'].unique())


Các ngày nhà hàng mở cửa ['Sun' 'Sat' 'Thur' 'Fri']


### 📑 Question 8 (EXPRESSION - 5 pts)
Write one line of code to get the list of **unique values** of the column ```time```.

**Output should look like this**

array(['Dinner', 'Lunch'], dtype=object)

In [37]:
# Your answer/code here:
df['time'].unique()

array(['Dinner', 'Lunch'], dtype=object)

### 📑 Question 9 (EXPRESSION - 5 pts)
Write one line of code to get the range of table size. 

Range means the **difference between the max and the min**.

**Output should be a number**

In [38]:
# Your answer/code here:
size_r = print(df['size'].max() - df['size'].min())


5


### 📑 Question 10 (MULTICHOICE_SINGLE - 5 pts)
How many men/women paid for the dish?

- A Male: 157, Female: 89

- B Male: 85, Female: 156

- C Male: 157, Female: 87

- D Male: 142, Female: 15

In [40]:
# Your answer/code here:

count_male_paid = ((df['total_bill'] > 0) & (df['sex'] == 'Male')).sum()
count_female_paid = ((df['total_bill'] > 0) & (df['sex'] == 'Female')).sum()

print(f'⌚ Số người nam đã thanh toán là: {count_male_paid}')
print(f'✂ Số người nữ đã thanh toán là: {count_female_paid}')

⌚ Số người nam đã thanh toán là: 157
✂ Số người nữ đã thanh toán là: 87


### 📑 Question 11 (MULTICHOICE_SINGLE - 5 pts)
What is the percentage of number of smoker table over the total?

- A 38%

- B 40%

- C 2%

- D 62%

In [41]:
df.shape[0]

244

In [42]:
# Your answer/code here:
per_smoker_table = 100 * sum(df['smoker'] == 'Yes') / df.shape[0]
print(f"❌ Tỉ lệ bàn có hút thuốc là: {per_smoker_table:.2f}%")

❌ Tỉ lệ bàn có hút thuốc là: 38.11%


In [43]:
# Your answer/code here:
per_smoker_table = round((df[df['smoker'] == 'Yes'].shape[0] / df.shape[0])*100,2)
print(f"❌ Tỉ lệ bàn có hút thuốc là: {per_smoker_table:.2f}%")

❌ Tỉ lệ bàn có hút thuốc là: 38.11%


### 📑 Question 12 (MULTICHOICE_SINGLE - 5 pts)
What is the percentage of revenue from the smoker tables over the total?

- A 38%

- B 40%

- C 2%

- D 62%

In [ ]:
df[df['smoker'] == 'Yes']   # Lọc lấy riêng smoker

total_bill   tip     sex smoker  day    time  size
56        38.01  3.00    Male    Yes  Sat  Dinner     4
58        11.24  1.76    Male    Yes  Sat  Dinner     2
60        20.29  3.21    Male    Yes  Sat  Dinner     2
61        13.81  2.00    Male    Yes  Sat  Dinner     2
62        11.02  1.98    Male    Yes  Sat  Dinner     2
..          ...   ...     ...    ...  ...     ...   ...
234       15.53  3.00    Male    Yes  Sat  Dinner     2
236       12.60  1.00    Male    Yes  Sat  Dinner     2
237       32.83  1.17    Male    Yes  Sat  Dinner     2
240       27.18  2.00  Female    Yes  Sat  Dinner     2
241       22.67  2.00    Male    Yes  Sat  Dinner     2

[93 rows x 7 columns]

In [48]:
df[df['smoker'] == 'Yes']['total_bill'].sum()   # Doanh thu từ smoker


1930.3400000000001

In [49]:
df['total_bill'].sum()      # Tổng doanh thu của nhà hàng

4827.77

In [104]:
# Your answer/code here:
per_rev_smoker = 100 * df[df['smoker'] == 'Yes']['total_bill'].sum() / df['total_bill'].sum()
print(f"🚀 Tỉ lệ doanh thu của bàn có hút thuốc là: {per_rev_smoker:.2f}%")

🚀 Tỉ lệ doanh thu của bàn có hút thuốc là: 39.98%


In [105]:
# Your answer/code here:

total_rev_smoker = df.loc[df['smoker'] == 'Yes', 'total_bill'].sum()    # Tính tổng doanh thu từ người hút thuốc
total_rev_rest = df['total_bill'].sum()                                 # Tính tổng doanh thu của nhà hàng
per_rev_smoker = (total_rev_smoker/total_rev_rest)*100
print(f'{per_rev_smoker:.2f}%')

39.98%


### 📑 Question 14 (MULTICHOICE_SINGLE - 5 pts)
Between smoker and non-smoker, who tips more according the percentage of the total tip on total bill.

- A Non-smoker

- B Smoker

In [106]:
# Your answer/code here:
total_tip_smoker = df.loc[df['smoker'] == 'Yes', 'tip'].sum()    # Tính tổng tip từ người hút thuốc
total_tip_rest = df['tip'].sum()                                 # Tính tổng tip của nhà hàng
per_tip_smoker = (total_tip_smoker/total_tip_rest)*100
per_tip_nonsmoker = 100 - per_tip_smoker
print(f'Percentage tip of smokers: {per_tip_smoker:.2f}%')
print(f'Percentage tip of nonsmokers: {per_tip_nonsmoker:.2f}%')

Percentage tip of smokers: 38.25%
Percentage tip of nonsmokers: 61.75%


### 📑 Question 15 (EXPRESSION - 5 pts)

Write one line of code to filter out the **"smoker"** tables on **Sunday**.

**Output should look like this**

|     |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|----:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
| 164 |        17.51 |  3    | Female | Yes      | Sun   | Dinner |      2 |
| 172 |         7.25 |  5.15 | Male   | Yes      | Sun   | Dinner |      2 |
| 173 |        31.85 |  3.18 | Male   | Yes      | Sun   | Dinner |      2 |
| 174 |        16.82 |  4    | Male   | Yes      | Sun   | Dinner |      2 |
| 175 |        32.9  |  3.11 | Male   | Yes      | Sun   | Dinner |      2 |
| 176 |        17.89 |  2    | Male   | Yes      | Sun   | Dinner |      2 |
| 177 |        14.48 |  2    | Male   | Yes      | Sun   | Dinner |      2 |
| 178 |         9.6  |  4    | Female | Yes      | Sun   | Dinner |      2 |
| 179 |        34.63 |  3.55 | Male   | Yes      | Sun   | Dinner |      2 |
| 180 |        34.65 |  3.68 | Male   | Yes      | Sun   | Dinner |      4 |
| 181 |        23.33 |  5.65 | Male   | Yes      | Sun   | Dinner |      2 |
| 182 |        45.35 |  3.5  | Male   | Yes      | Sun   | Dinner |      3 |
| 183 |        23.17 |  6.5  | Male   | Yes      | Sun   | Dinner |      4 |
| 184 |        40.55 |  3    | Male   | Yes      | Sun   | Dinner |      2 |
| 186 |        20.9  |  3.5  | Female | Yes      | Sun   | Dinner |      3 |
| 187 |        30.46 |  2    | Male   | Yes      | Sun   | Dinner |      5 |
| 188 |        18.15 |  3.5  | Female | Yes      | Sun   | Dinner |      3 |
| 189 |        23.1  |  4    | Male   | Yes      | Sun   | Dinner |      3 |
| 190 |        15.69 |  1.5  | Male   | Yes      | Sun   | Dinner |      2 |

In [107]:
# Your answer/code here:
smoker_sun = df[(df['smoker'] == 'Yes') & (df['day'] == 'Sun')]
smoker_sun

total_bill   tip     sex smoker  day    time  size
164       17.51  3.00  Female    Yes  Sun  Dinner     2
172        7.25  5.15    Male    Yes  Sun  Dinner     2
173       31.85  3.18    Male    Yes  Sun  Dinner     2
174       16.82  4.00    Male    Yes  Sun  Dinner     2
175       32.90  3.11    Male    Yes  Sun  Dinner     2
176       17.89  2.00    Male    Yes  Sun  Dinner     2
177       14.48  2.00    Male    Yes  Sun  Dinner     2
178        9.60  4.00  Female    Yes  Sun  Dinner     2
179       34.63  3.55    Male    Yes  Sun  Dinner     2
180       34.65  3.68    Male    Yes  Sun  Dinner     4
181       23.33  5.65    Male    Yes  Sun  Dinner     2
182       45.35  3.50    Male    Yes  Sun  Dinner     3
183       23.17  6.50    Male    Yes  Sun  Dinner     4
184       40.55  3.00    Male    Yes  Sun  Dinner     2
186       20.90  3.50  Female    Yes  Sun  Dinner     3
187       30.46  2.00    Male    Yes  Sun  Dinner     5
188       18.15  3.50  Female    Yes  Sun  Dinner     3
189       23.10  4.00    Male    Yes  Sun  Dinner     3
190       15.69  1.50    Male    Yes  Sun  Dinner     2

### 📑 Question 16 (MULTICHOICE_SINGLE - 5 pts)
How many tables belong to the smoker, at lunch?

- A 20

- B 21 

- C 22

- D 23

In [110]:
# Your answer/code here:
print("Số bàn của smoker vào buổi Lunch là:", df[(df['smoker'] == 'Yes') & (df['time'] == 'Lunch')].shape[0])


Số bàn của smoker vào buổi Lunch là: 23


In [109]:
# Your answer/code here:
# Lọc ra các dòng có smoker = 'Yes' và time = 'Lunch'
smoker_lunch = df[(df['smoker'] == 'Yes') & (df['time'] == 'Lunch')]

# Đếm số lượng dòng = số bàn
count = len(smoker_lunch)

print("Số bàn của smoker vào buổi Lunch là:", count)

Số bàn của smoker vào buổi Lunch là: 23


### 📑 Question 17 (EXPRESSION - 10 pts)

Write one line of code to compute the **average tip** amount of men who smoke and dine on **Friday's dinner**.


**Output should be a number**

In [111]:
# Your answer/code here:
avg_tip = df[(df['sex'] == 'Male') & (df['smoker'] == 'Yes') & (df['day'] == 'Fri') & (df['time'] == 'Dinner')]['tip'].mean()
avg_tip

3.246

### 📑 Question 18 (EXPRESSION - 5 pts)
Write one line of code to compute the **sum** of total bills of the **male** customers who dine in any day that's **not Friday**. 

**Output should be a number**

In [112]:
# Your answer/code here:
total_sum = df[(df['sex'] == 'Male') & (df['day'] != 'Fri')]['total_bill'].sum()
total_sum

3058.25

### 📑 Question 19 (EXPRESSION - 5 pts)
Write one line of code to filter the tables with tip amount **larger** than the **average** tip amount. 

In [113]:
# Your answer/code here:
above_avg_tips = df[df['tip'] > avg_tip]
above_avg_tips

total_bill   tip     sex smoker  day    time  size
2         21.01  3.50    Male     No  Sun  Dinner     3
3         23.68  3.31    Male     No  Sun  Dinner     2
4         24.59  3.61  Female     No  Sun  Dinner     4
5         25.29  4.71    Male     No  Sun  Dinner     4
11        35.26  5.00  Female     No  Sun  Dinner     4
..          ...   ...     ...    ...  ...     ...   ...
214       28.17  6.50  Female    Yes  Sat  Dinner     3
221       13.42  3.48  Female    Yes  Fri   Lunch     2
232       11.61  3.39    Male     No  Sat  Dinner     2
238       35.83  4.67  Female     No  Sat  Dinner     3
239       29.03  5.92    Male     No  Sat  Dinner     3

[83 rows x 7 columns]

### 📑 Question 20 (EXPRESSION - 5 pts)

- Task 1: load `time_hours.xlsx` table to `df_th` variable:

In [6]:
import pandas as pd
df_th = pd.read_excel(r'Data\time_hours.xlsx') # your code here

# show time_hours table:
df_th

time    hour
0  Dinner  19-21h
1   Lunch  10-12h

- Task 2: Get `hour` information for `df` dataframe from `df_th` by using Using `time` field (`pd.merge()`):

In [20]:
# your code here:

df_m = pd.merge(
    left=df,
    right=df_th,
    left_on='time',
    right_on='time',
    how='left'  
)
df_m.head(3)

total_bill   tip     sex smoker  day    time  size  hour_x  hour_y
0       16.99  1.01  Female     No  Sun  Dinner     2  19-21h  19-21h
1       10.34  1.66    Male     No  Sun  Dinner     3  19-21h  19-21h
2       21.01  3.50    Male     No  Sun  Dinner     3  19-21h  19-21h

✅ Giải thích từng phần:
1. `df['smoker'] == 'Yes'`
-   So sánh từng dòng trong cột 'smoker' với 'Yes'

-   Trả về một Series boolean: True nếu người đó là smoker, False nếu không.

2.`df[ ... ]`
-   Lọc DataFrame theo điều kiện bên trong dấu ngoặc vuông [ ... ]

-   Tức là chỉ lấy các dòng mà điều kiện là True → tức là các smoker

3. `['total_bill']`
-   Sau khi đã lọc ra các dòng, ta chọn riêng cột total_bill

4. `.sum()`
-   Tính tổng tiền total_bill từ các dòng còn lại (tức là của smoker)

✅ Kết quả cuối cùng:
-   Tính tổng doanh thu từ các khách hàng là smoker (smoker == 'Yes')

In [21]:
# your code here:
# Bỏ cột hour trong df trước khi merge:
df_m = pd.merge(df.drop(columns=['hour'], errors='ignore'),

    right=df_th,
    left_on='time',
    right_on='time',
    how='left'  
)
df_m.head(3)

total_bill   tip     sex smoker  day    time  size    hour
0       16.99  1.01  Female     No  Sun  Dinner     2  19-21h
1       10.34  1.66    Male     No  Sun  Dinner     3  19-21h
2       21.01  3.50    Male     No  Sun  Dinner     3  19-21h

In [22]:
# your code here:
# Bỏ cột hour trong df trước khi merge:
df_m = pd.merge(
    df.drop(columns=['hour'], errors='ignore'),
              df_th, on='time', how='left')

df_m.head(3)

total_bill   tip     sex smoker  day    time  size    hour
0       16.99  1.01  Female     No  Sun  Dinner     2  19-21h
1       10.34  1.66    Male     No  Sun  Dinner     3  19-21h
2       21.01  3.50    Male     No  Sun  Dinner     3  19-21h

In [23]:
# show random sample 5 rows:
df.sample(5)

total_bill   tip     sex smoker   day    time  size    hour
173       31.85  3.18    Male    Yes   Sun  Dinner     2  19-21h
148        9.78  1.73    Male     No  Thur   Lunch     2  10-12h
51        10.29  2.60  Female     No   Sun  Dinner     2  19-21h
107       25.21  4.29    Male    Yes   Sat  Dinner     2  19-21h
166       20.76  2.24    Male     No   Sun  Dinner     2  19-21h